### PR TITLE
Algebraic Data Types

### PR DESCRIPTION
**Pull request by MicaelLoberg**
This PR extends the Encore language with ADTs, which are cool and totally sweet.
There are some examples how to use them in the tests I've added.

### CODE REVIEW

**Comment by supercooldave** [comment]
Could be more tests, but given that Micael has moved on, I think this is worth keeping and fixing in the future.

**Comment by EliasC** [comment]
I have a review of this PR that I did while offline yesterday. Will post it here for posterity tomorrow.

**Comment by EliasC** [comment]
I wrote this review a day before it was merged. @MicaelLoberg is not expected to address these things (although if he wants to, please tell us below!). This is just the high-level comments. I will provide the code review later (for whomever is going to fix it).

- `data` should be a reserved keyword

- It's a bit annoying that a data constructor without arguments
  still needs parentheses:

  ```
  data MyBool
  case True() : MyBool
  case False : MyBool -- Parsing error: needs parentheses

  active class Main
    def main() : unit
      val t : MyBool = True -- fails: True is a function of type () -> MyBool
    end
  end
  ```

  This comes down to design decisions, and could probably be left
  as is for now.

- Providing a non existing data type for a case crashes the compiler:

  ```
  data MyBool
  case True() : MyBol
  ```
  > encorec: "adt.enc" (line 2, column 1): ADT "MyBol" not defined

  This looks like an attempt to do "type checking" in the
  desugarer, which is too early because we have no type
  information yet. The Haskell function `error` should _only_ be
  used for compiler bugs, i.e. things that should never happen.
  Blindly passing the type and let the type checker sort it out
  should be enough.

- Related to the above, using a trait or class type as the type of
  a case crashes the compiler:

  ```
  trait MyBool
  case True() : MyBool
  ```
  > encorec: "adt.enc" (line 10, column 1): ADT "Expr" not defined

- The parser allows trait compositions as the type of data cases,
  but this crashes the compiler:

  ```
  data MyBool
  data Foo
  case True() : MyBool + Foo -- Causes a crash!
  ```
  > encorec: src/ir/AST/Desugarer.hs:274:39-75: Non-exhaustive patterns in record update

  The simplest solution is of course to only allow a single
  ref-type (although it is exciting to think about what it would
  mean to have a case that belongs to two different datatypes!).

- Type inference for polymorphic data should be improved:

  ```
  data List[t]
  case Nil[t]() : List[t]
  case Cons[t](hd : t, tl : List[t]) : List[t]

  active class Main
    def main() : unit
      var l = Cons(1, Cons(2, Cons(3, Nil[int]())))
    end
  end
  ```

  Ideally, we wouldn't have to provide the type parameter to `Nil`
  I think you could leave this for future work.

- You can currently define methods in data declarations. It would
  make sense to also be able to `require` methods (which must then
  be provided by the cases):

  ```
  data Expr
    require eval() : int
  end

  case Add(l : Expr, r : Expr) : Expr
    def eval() : int
      this.l.eval() + this.r.eval()
    end
  end
  ```

  Since the data types are implemented as traits, this should more
  or less be a matter of telling the parser to allow (method)
  requirements as well.

- The parser is a bit eager to abort parsing a block data
  definition:

  ```
  data Foo
    def foo() : unit
      1 +
    end
  end
  ```
  > "adt.enc" (line 2, column 3):
  > unexpected 'd'

  The error should of course be on the next line, where is a
  missing expression.

- Related to the above, it isn't possible to override methods in
  cases:

  ```
  data Expr
    def eval() : int
      return -1
    end
  end

  case Add(l : Expr, r : Expr) : Expr
    def eval() : int
      this.l.eval() + this.r.eval()
    end
  end
  ```

  > "adt.enc" (line 1, column 1)
  > Duplicate declaration of method 'eval'
  > In class 'Add'

  (note also that the position in the error message is off)

  Manually desugaring into the corresponding trait/class pair does
  work, which suggests that the desugaring process for ADTs could
  be simplified:

  ```
  read trait Expr
    def eval() : int
      return -1
    end
  end

  class Add : Expr(l, r)
    val l : Expr
    val r : Expr
    def eval() : int
      this.l.eval() + this.r.eval()
    end
  end
  ```

File: src/ir/AST/AST.hs [hunk 1]
```diff
@@ -228,6 +230,26 @@ data ClassDecl = Class {
...
 
+data AdtDecl = ADT {
+  ameta        :: Meta AdtDecl,
+  aname        :: Type,
+  aconstructor :: [AdtConstructor],
+  amethods     :: [MethodDecl],
+  identity     :: String
+} deriving (Show)
+
+data AdtConstructor = ADTcons {
```
Comment (EliasC) [inline src/ir/AST/AST.hs:241]: I would call these `AdtCase` and `ADTCase` respectively.

File: src/ir/AST/AST.hs [hunk 2]
```diff
@@ -228,6 +230,26 @@ data ClassDecl = Class {
...
+  acmeta         :: Meta AdtConstructor,
+  acname         :: Type,
+  acfields       :: [ParamDecl],
+  acomposition   :: TraitComposition,
```
Comment (EliasC) [inline src/ir/AST/AST.hs:245]: Unless we want the ability to write compositions here, this should most likely be a `Type`.

File: src/ir/AST/AST.hs [hunk 3]
```diff
@@ -228,6 +230,26 @@ data ClassDecl = Class {
...
+  acmethods      :: [MethodDecl],
+  parentIdentity :: String
```
Comment (EliasC) [inline src/ir/AST/AST.hs:247]: Is this the name of the `data` declaration? If so, it should be a `Type`.

File: src/ir/AST/AST.hs [hunk 4]
```diff
@@ -255,6 +277,20 @@ instance HasMeta ClassDecl where
...
+    setType _ _ =
+        error "AST.hs: Cannot set the type of an ADT"
+    showWithKind ADT{aname} = "data '" ++ getId aname ++ "'"
+
+instance HasMeta AdtConstructor where
+    getMeta = acmeta
+    setMeta ac m = ac{acmeta = m}
+    setType _ _ =
+        error "AST.hs: Cannot set the type of an ADT"
+    showWithKind ADTcons{acname} = "data '" ++ getId acname ++ "'" --TODO: ?
```
Comment (EliasC) [inline src/ir/AST/AST.hs:292]: It doesn't make sense to print both `data` and `case` declarations as "data". In other words, the TODO is well placed =)

File: src/types/Types.hs [hunk 5]
```diff
@@ -809,6 +883,19 @@ isRefType ty
 isUnresolved Type{inner = Unresolved{}} = True
 isUnresolved _ = False
 
+getAdtTag Type{inner = ClassType{adtTag}} = adtTag
+getAdtTag _ = 0
```
Comment (EliasC) [inline src/types/Types.hs:887]: If this function is only supposed to be called on class types, it's better to throw an error here!

File: src/types/Types.hs [hunk 6]
```diff
@@ -794,6 +866,8 @@ isRefAtomType Type{inner = Unresolved {}} = True
 isRefAtomType Type{inner = TraitType {}} = True
 isRefAtomType Type{inner = AbstractTraitType {}} = True
 isRefAtomType Type{inner = ClassType {}} = True
+isRefAtomType Type{inner = AdtType {}} = True -- TODO: Ask someone about this
+isRefAtomType Type{inner = AdtConsType {}} = True -- TODO: Ask someone about this
```
Comment (EliasC) [inline src/types/Types.hs:870]: Did you ask someone? =)

File: src/types/Types.hs [hunk 7]
```diff
@@ -502,6 +517,8 @@ showWithKind ty = kind (inner ty) ++ " " ++ show ty
                                          else if isSharedSingleType (typ ty)
                                          then "shared class type"
                                          else "class type"
+    kind AdtType{}                     = "ADT type"
```
Comment (EliasC) [inline src/types/Types.hs:520]: Shouldn't this be "abstract data type"?

File: src/types/Typechecker/Typechecker.hs [hunk 8]
```diff
@@ -1100,6 +1115,75 @@ instance Checkable Expr where
...
+        doGetAdtPatternVars pt fcall@(FunctionCall {qname, args = []}) = do
+          {-header <- findMethod pt (qnlocal qname)-}
+          {-let hType = htype header-}
+              {-extractedType = getResultType hType-}
+          {-unless (isUnitType extractedType) $ do-}
+                 {-let expectedLength = if isTupleType extractedType-}
+                                      {-then length (getArgTypes extractedType)-}
+                                      {-else 1-}
+                 {-tcError $ PatternArityMismatchError (qnlocal qname)-}
+                           {-expectedLength 0-}
```
Comment (EliasC) [inline src/types/Typechecker/Typechecker.hs:1143]: Dead code

File: src/types/Typechecker/Typechecker.hs [hunk 9]
```diff
@@ -1100,6 +1115,75 @@ instance Checkable Expr where
...
+        doGetAdtPatternVars pt va@(VarAccess {qname}) = do
+          when (isThisAccess va) $
+            tcError ThisReassignmentError
+          return [(qnlocal qname, pt)]
+
+        {-doGetAdtPatternVars pt mcp@(MaybeValue{mdt = JustData {e}})-}
+            {-| isMaybeType pt =-}
+                {-let innerType = getResultType pt-}
+                {-in getAdtPatternVars innerType e-}
+            {-| otherwise = tcError $ PatternTypeMismatchError mcp pt-}
```
Comment (EliasC) [inline src/types/Typechecker/Typechecker.hs:1132]: Dead code

File: src/parser/Parser/Parser.hs [hunk 10]
```diff
@@ -750,6 +758,131 @@ partitionClassAttributes = partitionClassAttributes' [] []
...
+--
+--    lineDecl ameta = do
+--      name <- lookAhead upperChar >> identifier
+--      params <- optionalTypeParameters
+--      return ADT{ameta
+--                ,aname = setRefNamespace emptyNamespace $ adtType name params
+--                ,aconstructor = []
+--                ,amethods = []
+--                ,identity = name
+--                }
```
Comment (EliasC) [inline src/parser/Parser/Parser.hs:831]: Lots of dead code!

File: src/ir/AST/AST.hs [hunk 11]
```diff
@@ -586,6 +622,12 @@ data Expr = Skip {emeta :: Meta Expr}
                          args :: Arguments}
           | Optional {emeta :: Meta Expr,
                       optTag :: OptionalPathComponent}
+          | AdtExtractorPattern {emeta :: Meta Expr,
+                                 ty :: Type,
+                                 name :: Name,
+                                 arg :: Expr,
+                                 fieldNames :: [String],
+                                 adtClassDecl :: ClassDecl}
```
Comment (EliasC) [inline src/ir/AST/AST.hs:630]: Does this have to be a `ClassDecl`, or could it be a type that is looked up when needed?

File: src/ir/AST/AST.hs [hunk 12]
```diff
@@ -228,6 +230,26 @@ data ClassDecl = Class {
   cmethods    :: [MethodDecl]
 } deriving (Show)
 
+data AdtDecl = ADT {
+  ameta        :: Meta AdtDecl,
+  aname        :: Type,
+  aconstructor :: [AdtConstructor],
+  amethods     :: [MethodDecl],
+  identity     :: String
```
Comment (EliasC) [inline src/ir/AST/AST.hs:238]: Isn't this information already in `aname`?

**Review by EliasC — COMMENTED** [review]
A few comments on the code.

**Pull request merged**